src/Orleans/Core/GrainFactory.cs

```diff
@@ -167,6 +179,50 @@ internal Task<TGrainObserverInterface> CreateObjectReference<TGrainObserverInter
             return CreateObjectReferenceImpl<TGrainObserverInterface>(obj);
         }
 
+        private static void RegisterGrainTypes(Type grainInterfaceType)
+        {
+            // Do a one-time discovery of all the types used by this grain interface and if there is a serializer
+            // for it, make sure the type is registered with the serialization manager.
+            GrainRegistrationDictionary.GetOrAdd(
+                grainInterfaceType,
+                type =>
+                {
+                    if (type.IsConstructedGenericType)
+                    {
+                        foreach (var typeParam in type.GetGenericArguments().Where(SerializationManager.HasExternalSerializer))
+                        {
+                            SerializationManager.Register(typeParam);
+                        }
+                    }
+
+                    foreach (var paramType in 
+                        type.GetMethods(BindingFlags.Instance | BindingFlags.Public)
+                        .SelectMany(
+                            mi =>
+                            {
+                                var parameterTypes = mi.GetParameters().Select(pi => pi.ParameterType);
+                                if (mi.ReturnType == typeof(void) || mi.ReturnType == typeof(Task))
+                                {
+                                    return parameterTypes;
+                                }
+
+                                if (mi.ReturnType.IsGenericType && mi.ReturnType.GetGenericTypeDefinition() == typeof(Task<>))
+                                {
+                                    return parameterTypes.Concat(mi.ReturnType.GetGenericArguments());
+                                }
+
+                                return parameterTypes.Concat(new[] { mi.ReturnType });
+                            })
+                        .Distinct()
+                        .Where(SerializationManager.HasExternalSerializer))
+                    {
+                        SerializationManager.Register(paramType);
```

Comment:
After more investigation, the Assembly Processor does a brute force scan of all assemblies that get loaded to find serialization info.  A minor change to that logic should be all that is necessary to fix this and handle the derived types issue as well.

Sorry for the churn.



The current mechanism is a bit verbose as well.